feat: add sbom generation

wagoodman opened the pull request:
This PR adds a new `sboms` section to generate SBOMs for one or more generated artifacts. Example usage:
```
sboms:
  -
    id: binary-sbom
    artifacts: binary

```
See doc updates in this PR for more detailed configuration explanation. With this approach SBOMs generated can be checksummed and referenced under the `signs` section.

_(This is a follow up from https://github.com/goreleaser/goreleaser/issues/2597#issuecomment-957842629 .)_

caarlos0 commented on internal/pipe/pipe.go:
```diff
@@ -25,6 +25,9 @@ var ErrSkipSignEnabled = Skip("artifact signing is disabled")
 // It means that the part of a Piper that validates some things was not run.
 var ErrSkipValidateEnabled = Skip("validation is disabled")
 
+// ErrSkipSBOMEnabled happens if --skip-sbom is set.
+var ErrSkipSBOMEnabled = Skip("artifact cataloging is disabled")
```
not being used anywhere...

wagoodman replied:
correct! With the addition of https://github.com/goreleaser/goreleaser/pull/2648#discussion_r747133802 now it should be 👍

caarlos0 replied:
seems like still unused... 

in any case, I don't think its really needed... the `Skip` method in the pipe handles it :)

caarlos0 commented on internal/pipe/sbom/sbom.go:
```diff
@@ -0,0 +1,222 @@
+type Pipe struct{}
+
+func (Pipe) String() string                 { return "cataloging artifacts" }
+func (Pipe) Skip(ctx *context.Context) bool { return len(ctx.Config.SBOMs) == 0 }
```
```suggestion
func (Pipe) Skip(ctx *context.Context) bool { return ctx.SkipSBOMCataloging || len(ctx.Config.SBOMs) == 0 }
```

caarlos0 commented on internal/pipe/sign/sign.go:
```diff
@@ -77,11 +77,12 @@ func (Pipe) Run(ctx *context.Context) error {
 				))
 			case "archive":
 				filters = append(filters, artifact.ByType(artifact.UploadableArchive))
-			case "binary":
-				filters = append(filters, artifact.ByType(artifact.UploadableBinary))
```
likely removed by mistake

wagoodman replied:
indeed it was! I'll restore

caarlos0 commented on internal/pipeline/pipeline.go:
```diff
@@ -4,6 +4,8 @@ package pipeline
 	"fmt"
 
+	"github.com/goreleaser/goreleaser/internal/pipe/sbom"
+
```
```suggestion
```

caarlos0 commented on pkg/defaults/defaults.go:
```diff
@@ -5,6 +5,8 @@ package defaults
 	"fmt"
 
+	"github.com/goreleaser/goreleaser/internal/pipe/sbom"
+
```
```suggestion
```

caarlos0 commented on internal/pipe/sbom/sbom.go:
```diff
@@ -0,0 +1,222 @@
+			cfg.SBOMs = []string{sbom}
+		}
+		if len(cfg.Args) == 0 {
+			cfg.Args = []string{"--file", "{{ .Dist }}/$sbom0", "--output", "spdx-json", "$artifact"}
```
on the sign pipe prefix the file with the dist when setting the env... I would say we follow the same path here so both usages/impls look similar...

caarlos0 reviewed (commented):
I like this!

caarlos0 commented on cmd/release.go:
```diff
@@ -139,6 +141,7 @@ func setupReleaseContext(ctx *context.Context, options releaseOpts) *context.Con
 	ctx.SkipAnnounce = ctx.Snapshot || options.skipPublish || options.skipAnnounce
 	ctx.SkipValidate = ctx.Snapshot || options.skipValidate
 	ctx.SkipSign = options.skipSign
+	ctx.SkipSBOMCataloging = options.skipSBOMCataloging
```
should `--skip-validate` infers `--skip-sbom`?

wagoodman replied:
My take is that skipping validation shouldn't imply skipping SBOM generation. That is, cataloging artifacts isn't a form of validation and indeed creates artifacts of their own.

caarlos0 replied:
hmm, yeah, that makes sense...

caarlos0 commented on internal/artifact/artifact.go:
```diff
@@ -56,6 +56,8 @@ const (
+	SBOM
 )
 
 func (t Type) String() string {
```
missing sbom bellow here and in the test

caarlos0 commented on internal/pipe/sbom/sbom.go:
```diff
@@ -0,0 +1,228 @@
+
+		var paths []string
+		for idx, sbom := range cfg.SBOMs {
+			path, err := templater.Apply(filepath.Join("{{ .Dist }}", expand(sbom, env)))
```
could be 
```suggestion
			path, err := templater.Apply(filepath.Join(ctx.Dist, expand(sbom, env)))
```

caarlos0 commented on internal/tmpl/tmpl.go:
```diff
@@ -27,6 +27,7 @@ type Fields map[string]interface{}
 const (
 	// general keys.
 	projectName     = "ProjectName"
+	dist            = "Dist"
```
I think this is not needed... our internal code can use `ctx.Config.Dist` and user-configs should be scoped to dist already...

wagoodman replied:
Circling back to this --I added `Dist` as an option to allow for tool arguments that have to deal with outputting files to allow them to specify the dist directory explicitly:
```
...
-
  Cmd: my-tool
  Args:
    - --file
    - "{{ .Dist }}/some.sbom"
...
```

wagoodman replied:
After more thought I decided to go the route of changing the cwd of the SBOM tool to `Dist` and removing `Dist` from templating. This way all cataloging tools are oriented towards what goreleaser generates instead of another location. An additional pro is that it makes it harder to misuse the `$sbom#` variables (you don't need to remember to prefix `{{ .Dist }}/` to all paths). 

The downside of this is that if tooling depends on configuration in the root of the repo this will now need to be explicitly passed in as an option.

caarlos0 replied:
yes, that was my idea as well!

sorry it took so long to reply, crazy week :P

caarlos0 reviewed (commented):
Its looking good!

I think we now need:
- more tests
- docs
- `task fmt`/`task lint` to make ci happy

let me know how I can help :)

wagoodman commented:
Righto @caarlos0 , thanks for the extra review (the comments are helpful ❤️ ). I'll adjust the current branch and start adding docs / tests (it seems like the direction has been settled)... when I'm done I'll take it out of draft (🎉 ).

caarlos0 commented:
amazing, let me know if you need any help in the meantime (I'm on our dicord and also gophers slack as well)

wagoodman commented:
@caarlos0 I think this one is ready for review 🎉 , when CI gets kicked off I'll come back and see what the results were 🤞

caarlos0 reviewed (commented):
just a couple of comments, but, looking good!

I also played with it locally, and it works like a charm!

I would say we can enable this for archives on our own `.goreleaser.yaml` as well, dogfooding and all. 

Probably sign the sboms as well? Would generate a lot of files, but should be ok I think 🙇‍♂️

caarlos0 commented on www/mkdocs.yml:
```diff
@@ -91,6 +91,8 @@ nav:
     - customization/snapcraft.md
     - customization/docker.md
     - customization/docker_manifest.md
+  - Software Bill of Materials:
+      - Cataloging artifacts: customization/sbom.md
```
this could probably be just:

```suggestion
  - customization/sbom.md
```

wagoodman commented on www/docs/customization/sbom.md:
````diff
@@ -0,0 +1,100 @@
+```yaml
+# .goreleaser.yml
+sboms:
+  - artifacts: binary
````
yeah, that makes a lot of sense. I'll change the default artifacts value and update the docs example

wagoodman commented:
@caarlos0 thanks! I added the new `sboms` section to the local `.goreleaser.yaml` and also added a new `signs` config. You certainly won't have a shortage of files on releases!

wagoodman commented:
I'm going to add a few more tests under `sbom_test.go` to get coverage up and I realized I'm missing tests for `sbom` artifacts under `signs_test.go`, so I'll add those too.

caarlos0 commented on www/docs/customization/sbom.md:
```diff
@@ -0,0 +1,102 @@
+
+## Limitations
+
+Container images generated by Goreleaser are not available to be cataloged by the SBOM tool.
```
As far as I tested, it don't work with macos universal binaries as well... so its probably worth listing it here too..

wagoodman replied:
I think I can get that added to syft and released shortly. I'll update here if it ends up being too large of a lift / too long.

caarlos0 replied:
ah, that sounds good too!

wagoodman replied:
Syft release v0.32.0 now has mac multi-architecture support! (updated the pipeline to use the latest release too) 🎉

caarlos0 replied:
niceee

caarlos0 commented:
Thanks @wagoodman!

wagoodman commented:
Righto! Twas fun to develop --Thanks for the conversations and reviews 🙌!
